### PR TITLE
[core] Hybrid scheduling policy.

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This PR introduces a new scheduling policy which is a hybrid of a pack and round robin policy. Description from the doc string:

```
/// This scheduling policy was designed with the following assumptions in mind:
///   1. Scheduling a task on a new node incurs a cold start penalty (warming the worker
///   pool).
///   2. Past a certain utilization threshold, a big noisy neighbor problem occurs (caused
///   by object spilling).
///   3. Locality is helpful, but generally outweighed by (1) and (2).
///
/// In order to solve these problems, we use the following scheduling policy.
///   1. Generate a traversal.
///   2. Run a priority scheduler.
///
/// A node's priorities are determined by the following factors:
///   * Always skip infeasible nodes
///   * Always prefer available nodes over feasible nodes.
///   * Break ties in available/feasible by critical resource utilization.
///   * Critical resource utilization below a threshold should be trucnated to 0.
///
/// The traversal order should:
///   * Prioritize the local node above all others.
///   * All other nodes should have a globally fixed priority across the cluster.
///
/// We call this a hybrid policy because below the threshold, the traversal and truncation
/// properties will lead to packing of nodes. Above the threshold, the policy will act
/// like a traditional weighted round robin.
```

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:175]: Can we have this only for CPU and object_store_memory? It can have unexpected effects on custom resources.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:347]: Comment

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:358]:
```suggestion
    std::sort(round.begin() + 1, round.end());
```
?

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:365]: Can you add unit tests for this method? It would be best to split it out into a standalone function that can be tested independently of the entire class.

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:383]: It's quite confusing to follow this, please clean it up and add more comments.

**Review by ericl — CHANGES_REQUESTED** [review]

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:218]:
```suggestion
int64_t ClusterResourceScheduler::GetBestSchedulableNodeSimpleBinPack(
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.h
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.h:155]: Please reword the algorithm in simpler terms (so a reader unfamiliar with that book can understand how it is working).

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:175]: Hmmm good point... I can imagine similar situations for CPU and memory though.

How about we do all predefined resources and no custom resources?

**Comment by ericl** [comment]
We can also include memory, but I think we should exclude GPU. Bin packing
for GPU is always preferable since they are so expensive.

**Comment by wuisawesome** [comment]
> Bin packing for GPU is always preferable since they are so expensive.

The GPU is in the cluster either way though right? Is this a heuristic to help with scaling down?

**Comment by ericl** [comment]
Yes. Also, GPUs don't really get oversubscribed in the way cpu and memory
does, so it doesn't make sense to be trying to spread.

**Comment by rkooo567** [comment]
Is it still WIP btw?

File: src/ray/common/ray_config_def.h
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:114]: What's happening if both this and `scheduler_loadbalance_spillback` are true?

File: src/ray/raylet/scheduling/scheduling_policy.h
Comment (rkooo567) [inline src/ray/raylet/scheduling/scheduling_policy.h:41]: Can you add a TODO to move other policies to this file? (load balancing & simple bin packing)

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:171]: Is it every happened? Why don't we just add a check here?

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:183]: Maybe just use std::min(highest, utilization)?

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:167]: Do we have unit tests for this function?

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:189]: Same here. Is this condition ever invoked? Why don't we just add a check?

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_data.cc:219]: This looks almost identical to `IsAvailable` except that it uses total instead of available. Any good way to reduce code duplication?

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_data.h:170]: Do we still need `IsFeasible` in the `cluster_resource_scheduler` after this? (except for the legacy logic)

**Review by rkooo567 — COMMENTED** [review]
Can we add one python test? I think we can make it something like this; 
1. 3 nodes cluster (0 cpu head node)
2. Each node has 100MB object store memory & 2 cpus
3. Make a task that returns 55MB object.
4. Create 3 tasks
5. Make sure all nodes has 55MB object memory usage.

File: src/ray/common/ray_config_def.h
Comment (rkooo567) [reply src/ray/common/ray_config_def.h:114]: Looks like `scheduler_loadbalance_spillback ` is ignored. Can you add a TODO comment to clean this up?

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:171]: yeah this actually happens. check was inspired by real events :p (most of the unit tests actually trigger it)

File: src/ray/raylet/scheduling/scheduling_policy.h
Comment (wuisawesome) [reply src/ray/raylet/scheduling/scheduling_policy.h:41]: Oh I was actually thinking we would delete those, but sure

File: src/ray/common/ray_config_def.h
Comment (wuisawesome) [reply src/ray/common/ray_config_def.h:114]: This flag takes precedence over the other one.

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (rkooo567) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:171]: That's interesting... Why is that? I thought the first N entries are reserved for predefined resources..

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_data.h:170]: I think it's only used by the legacy scheduler.

**Review by rkooo567 — APPROVED** [review]
Btw I will still wait for Eric's approval, but it LGTM if you add a python test.

File: src/ray/raylet/scheduling/cluster_resource_data.h
Comment (rkooo567) [reply src/ray/raylet/scheduling/cluster_resource_data.h:170]: Maybe just adding TODO comments on those functions so that we can easily clean up later?

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:171]: Yeah but the original data structure is actually a vector, and it's not dynamically resized at initialization time. This means that for some time, a task req of {"CPU": 1} will have `predefined_resources.size() == 1` until some code comes along and resizes the vector.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Failing C++ tests

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (clarkzinzow) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:334]: Supernit:
```suggestion
    // This is an actor which requires no resources.
    // Pick a random node to to avoid scheduling all actors on the local node.
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (clarkzinzow) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:344]: Nit: This should be equivalent and is a lot clearer to me.
```suggestion
      best_node = std::next(nodes_.begin(), idx).first;
```

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (clarkzinzow) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:171]: That's a bit of a funky data model... Given that predefined resources are static by definition, wouldn't it be a lot cleaner if `predefined_resources` was a static array whose elements may be unset? That would still allow for enum-based indexing (which I like), but would get rid of all of the dynamic resizing and size checks, and should make it far less brittle to hard-to-catch writer bugs around improper resizing. If that might make sense, this would obviously be a refactor that can wait for a future PR.

**Review by clarkzinzow — APPROVED** [review]
LGTM overall, all that I have are nits, no blockers for merging.

File: src/ray/raylet/scheduling/cluster_resource_data.cc
Comment (wuisawesome) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:170]: yeah we seem to have a convention of not using the underscore suffix for public members.

Comment (clarkzinzow) [reply src/ray/raylet/scheduling/cluster_resource_data.cc:170]: Ah, I didn't realize this was a public member! That's my bad.

**Comment by ericl** [comment]
This isn't compiling:
Use --sandbox_debug to see verbose messages from the sandbox
src/ray/raylet/scheduling/cluster_resource_scheduler.cc: In member function ‘int64_t ray::ClusterResourceScheduler::GetBestSchedulableNode(const TaskRequest&, bool, bool, int64_t*, bool*)’:
src/ray/raylet/scheduling/cluster_resource_scheduler.cc:338:50: error: ‘class absl::lts_2019_08_08::container_internal::raw_hash_set<absl::lts_2019_08_08::container_internal::FlatHashMapPolicy<long int, Node>, absl::lts_2019_08_08::hash_internal::Hash<long int>, std::equal_to<long int>, std::allocator<std::pair<const long int, Node> > >::iterator’ has no member named ‘first’
       best_node = std::next(nodes_.begin(), idx).first;
                                                  ^~~~~

**Comment by clarkzinzow** [comment]
@ericl That was my bad, my suggestion should have been
```cpp
best_node = std::next(nodes_.begin(), idx)->first;
```
🤦

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (clarkzinzow) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:338]: @wuisawesome In case you just want to commit a suggestion, my bad on this one.
```suggestion
      best_node = std::next(nodes_.begin(), idx)->first;
```

**Comment by wuisawesome** [comment]
@rkooo567 i'm going to remove the python test, it's way too flaky and exact scheduling behavior isn't guaranteed to users. we should just make sure the logic is correct (unit test) and the large scale behavior is correct (release test).

**Comment by wuisawesome** [comment]
I think the source of flakiness here is caused by the worker pool startup time, but it could involve the resource report updates too.

**Comment by wuisawesome** [comment]
Yeah i found a way to reproduce the flakiness locally, and i believe what's happening is that a bunch of tasks are being scheduled on the driver/first node, but because the warm pool hasn't started yet, they don't allocate resources, therefore more tasks are scheduled locally.

**Comment by rkooo567** [comment]
Hmm I see. Doesn't that mean this will also not work in real world due to the same issue? Btw, what does `locally` mean here? If you set 0 cpu head node, is it still reproducible?

**Comment by wuisawesome** [comment]
I didn't try with a 0 CPU head node, but there's nothing special about local scheduling anymore, so the same logic applies. The problem actually exists with the PACK scheduler too, but because of the packing semantics, you can't detect it from the outcome (just the number of spillbacks)

**Comment by rkooo567** [comment]
So, what's the conclusion here? Are we going to do the solution that you suggested in the slack?

**Comment by rkooo567** [comment]
Seems like Windows tests are timing out. `test_lease_request_leak`

**Comment by wuisawesome** [comment]
Here are some pretty pictures of scheduling throughput. They look as we would predict they would. 

hybrid_cdf
hybrid

**Comment by wuisawesome** [comment]
The remaining test failures seem unrelated. Lint fails due to SSL error (and passes in buildkite). The serve failure looks unrelated (serializing a dependency that changed?). I'm going to merge this now.

**Pull request merged**
